File: game/src/main/java/com/lentokonepeli/entities/Ground.java
package com.lentokonepeli.entities;

import com.lentokonepeli.Entity;
import com.lentokonepeli.EntityType;
import com.lentokonepeli.network.NetProp;
import com.lentokonepeli.network.NetType;

public class Ground extends Entity {

    private NetProp<Integer> x = new NetProp<>("x", NetType.i16);
    private NetProp<Integer> y = new NetProp<>("y", NetType.i16);
    private NetProp<Integer> width = new NetProp<>("width", NetType.i16);

    public Ground() {
        super(EntityType.GROUND);
        props.add(x);
        props.add(y);
        props.add(width);
    }
}

File: client/src/game/entities/ground.ts
import * as PIXI from "pixi.js";
import { EntityType } from "../../network/game/EntityType";
import { DrawLayer, Entity, TerrainType } from "../entity";
import { getTexture } from "../resources";
import { WATER_COLOR, WATER_DESERT_COLOR, WATER_HEIGHT } from "./water";

export interface GroundProps {
  x?: number;
  y?: number;
  width?: number;
  type?: number;
}

export class Ground extends Entity {
  x = 0;
  y = 0;
  width = 100;
  type = EntityType.GROUND;
  subType = TerrainType.NORMAL;

  container = new PIXI.Container();
  sprite = new PIXI.TilingSprite(getTexture("ground1.gif"));
  water = new PIXI.Graphics();

  constructor() {
    super();
    this.sprite.height = this.sprite.texture.height;

    this.container.sortableChildren = true;
    this.container.addChild(this.water);
    this.container.addChild(this.sprite);
    this.container.zIndex = DrawLayer.Ground;
  }

  getContainer(): PIXI.Container {
      return this.container;
  }

  redraw() {
    this.sprite.position.set(this.x, this.y);
    this.sprite.width = this.width;

    if (this.subType == TerrainType.DESERT) {
      this.sprite.texture = getTexture("groundDesert.gif");
    }

    const color = (this.subType == TerrainType.NORMAL) ? WATER_COLOR : WATER_DESERT_COLOR;
    this.water.beginFill(color);
    this.water.drawRect(this.x, this.y + this.sprite.height, this.width, WATER_HEIGHT);
    this.water.endFill();

    // TODO: change ground type

  }

  destroy() {}

}
